Java/src/com/couchbase/litecore/fleece/MArray.java

```diff
@@ -13,98 +13,189 @@
+        super.initInSlot(mv, parent, isMutable);
+        if (_baseArray != null)
+            throw new IllegalStateException("base array is not null.");
+
+        FLValue value = mv.getValue();
+        if (value != null) {
+            _baseArray  = value.asFLArray();
+            resize(_baseArray.count());
+        } else {
+            _baseArray = null;
+            resize(0);
+        }
     }
 
-    void initAsCopyOf(long hMDict, boolean isMutable) {
-        initAsCopyOf(_handle, hMDict, isMutable);
+    public void initAsCopyOf(MArray array, boolean isMutable) {
+        super.initAsCopyOf(array, isMutable);
+        _baseArray = array != null ? array.getBaseArray() : null;
+        _values = array != null ? new ArrayList<>(array._values) : new ArrayList<MValue>();
     }
 
+    /* Properties */
+
     public long count() {
-        return count(_handle);
+        return _values.size();
     }
 
-    public MValue get(int i) {
-        return new MValue(get(_handle, i));
-    }
 
-    public boolean set(int i, Object val) {
-        return set(_handle, i, val);
+    public FLArray getBaseArray() {
+        return _baseArray;
     }
 
-    public boolean insert(int i, Object val) {
-        return insert(_handle, i, val);
-    }
+    /* Public Methods */
 
-    public boolean append(Object val) {
-        return append(_handle, val);
-    }
+    /** Returns a reference to the MValue of the item at the given index.
+     If the index is out of range, returns an empty MValue. */
+    public MValue get(long index) {
+        if (index < 0 || index >= _values.size())
+            return MValue.EMPTY;
 
-    public boolean remove(int i) {
-        return remove(_handle, i, 1);
+        MValue value = _values.get((int)index);
+        if (value.isEmpty()) {
+            if (_baseArray != null) {
+                value = new MValue(_baseArray.get(index));
+                _values.set((int) index, value);
+            }
+        }
+        return value;
     }
 
-    public boolean remove(int i, int n) {
-        return remove(_handle, i, n);
-    }
+    public boolean set(long index, Object value) {
+        if (!isMutable())
+            throw new IllegalStateException("Cannot set items in a non-mutable MArray");
 
-    public boolean clear() {
-        return clear(_handle);
-    }
+        if (index < 0 || index >= count())
+            return false;
 
-    public void encodeTo(Encoder enc) {
-        encodeTo(_handle, enc._handle);
+        mutate();
+        _values.set((int)index, new MValue(value));
+        return true;
     }
 
-    @Override
-    protected void finalize() throws Throwable {
-        free();
-        super.finalize();
+    public boolean insert(long index, Object value) {
+        if (!isMutable())
+            throw new IllegalStateException("Cannot insert items in a non-mutable MArray");
+
+        if (index < 0 || index > count())
+            return false;
+
+        if (index < count())
```

Comment:
Is calling `populateValues()` once enough? I think it is not necessary to call the method always if `index < count()`.

Reply:
I followed the original MArray's implementation. Can you create a ticket to revise the implementation?

https://github.com/couchbaselabs/fleece/blob/260479eac5b274fae9642135978ab7b7d6edd3c5/Fleece/Mutable/MArray.hh#L93-L94

Reply:
https://github.com/couchbaselabs/fleece/issues/27